### DISCUSSION TITLE
Can't access `theme` & `showLineNumbers` directly in `pre` tag in `MDXComponents` while using MDX & Rehype Plugins on a Next.js blog?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [rehypejs/rehype#58] [Q&A] [answered]
I am creating a Next.js Blog & I'm facing a small issue.

I have MDX block like:

````mdx
```jsx theme=orange showLineNumbers=true
import React from 'react'

export function Counter({ initialCount = 0 }) {
	const [count, setCount] = React.useState(initialCount)
	return (
		<button type="button" onClick={() => setCount((prevCount) => prevCount + 1)}>
			{count}
		</button>
	)
}
```
````

I have a [Pre](https://github.com/deadcoder0904/better-code-blocks/blob/main/src/components/mdx/Pre.tsx) tag like:

```tsx
import React from 'react'

type PreProps = {
	children: React.ReactChild
	theme: string
	showLineNumbers: string
}

export const Pre = ({ children, theme, showLineNumbers }: PreProps) => {
	console.log('Pre -> start')
	console.log({ children, theme, showLineNumbers })
	console.log('Pre -> end')
	return (
		<pre
			className={`px-4 py-3 overflow-x-auto rounded-md ${
				theme ? `${theme}-theme` : 'bg-gray-900'
			} ${showLineNumbers ? 'line-numbers' : ''}`}
		>
			{children}
		</pre>
	)
}
```

However, I can't access `theme`, or `showLineNumbers` directly. I only get access to `children`.

And `children.props.theme` gives me an actual value which I want directly in `theme`. Similar thing with `children.props.showLineNumbers`.

I get the following when I use the `console.log` above:

```
{
  children: {
    '$$typeof': Symbol(react.element),
    type: { '$$typeof': Symbol(react.forward_ref), render: [Function] },
    key: null,
    ref: null,
    props: {
      parentName: 'pre',
      className: 'language-jsx',
      metastring: 'id=highlight',
      id: 'highlight',
      originalType: 'code',
      mdxType: 'code',
      children: [Array]
    },
    _owner: null,
    _store: {}
  },
  theme: undefined,
  showLineNumbers: undefined
}
```

This is probably because of some `rehype` plugins I use.

I have [highlight-word](https://github.com/deadcoder0904/better-code-blocks/blob/main/rehype/highlight-word.js) `rehype` plugin:

```tsx
const visit = require('unist-util-visit')
const hastToHtml = require('hast-util-to-html')
const unified = require('unified')
const parse = require('rehype-parse')

const CALLOUT = /__(.*?)__/g

module.exports = (code) => {
	const html = hastToHtml(code)
	const result = html.replace(CALLOUT, (_, text) => `<span class="highlight-word">${text}</span>`)
	const hast = unified().use(parse, { emitParseErrors: true, fragment: true }).parse(result)
	return hast.children
}
```

I have [highlight-line](https://github.com/deadcoder0904/better-code-blocks/blob/main/rehype/highlight-line.js) `rehype` plugin:

```tsx
const hastToHtml = require('hast-util-to-html')
const unified = require('unified')
const parse = require('rehype-parse')

const lineNumberify = function lineNumberify(ast, lineNum = 1) {
	let lineNumber = lineNum
	return ast.reduce(
		(result, node) => {
			if (node.type === 'text') {
				if (node.value.indexOf('\n') === -1) {
					node.lineNumber = lineNumber
					result.nodes.push(node)
					return result
				}

				const lines = node.value.split('\n')
				for (let i = 0; i < lines.length; i++) {
					if (i !== 0) ++lineNumber
					if (i === lines.length - 1 && lines[i].length === 0) continue
					result.nodes.push({
						type: 'text',
						value: i === lines.length - 1 ? lines[i] : `${lines[i]}\n`,
						lineNumber: lineNumber,
					})
				}

				result.lineNumber = lineNumber
				return result
			}

			if (node.children) {
				node.lineNumber = lineNumber
				const processed = lineNumberify(node.children, lineNumber)
				node.children = processed.nodes
				result.lineNumber = processed.lineNumber
				result.nodes.push(node)
				return result
			}

			result.nodes.push(node)
			return result
		},
		{ nodes: [], lineNumber: lineNumber }
	)
}

const wrapLines = function wrapLines(ast, linesToHighlight) {
	const highlightAll = linesToHighlight.length === 1 && linesToHighlight[0] === 0
	const allLines = Array.from(new Set(ast.map((x) => x.lineNumber)))
	let i = 0
	const wrapped = allLines.reduce((nodes, marker) => {
		const line = marker
		const children = []
		for (; i < ast.length; i++) {
			if (ast[i].lineNumber < line) {
				nodes.push(ast[i])
				continue
			}

			if (ast[i].lineNumber === line) {
				children.push(ast[i])
				continue
			}

			if (ast[i].lineNumber > line) {
				break
			}
		}

		nodes.push({
			type: 'element',
			tagName: 'div',
			properties: {
				dataLine: line,
				className: 'highlight-line',
				dataHighlighted: linesToHighlight.includes(line) || highlightAll ? 'true' : 'false',
			},
			children: children,
			lineNumber: line,
		})

		return nodes
	}, [])

	return wrapped
}

// https://github.com/gatsbyjs/gatsby/pull/26161/files
const MULTILINE_TOKEN_SPAN = /<span class="token ([^"]+)">[^<]*\n[^<]*<\/span>/g

const applyMultilineFix = function (ast) {
	// AST to HTML
	let html = hastToHtml(ast)

	// Fix JSX issue
	html = html.replace(MULTILINE_TOKEN_SPAN, (match, token) =>
		match.replace(/\n/g, `</span>\n<span class="token ${token}">`)
	)

	// HTML to AST
	const hast = unified().use(parse, { emitParseErrors: true, fragment: true }).parse(html)

	return hast.children
}

module.exports = function (ast, lines) {
	const formattedAst = applyMultilineFix(ast)
	const numbered = lineNumberify(formattedAst).nodes

	return wrapLines(numbered, lines)
}
```

I have a [highlight-code](https://github.com/deadcoder0904/better-code-blocks/blob/main/rehype/highlight-code.js) `rehype` plugin:

```tsx
const rangeParser = require('parse-numeric-range')
const visit = require('unist-util-visit')
const nodeToString = require('hast-util-to-string')
const refractor = require('refractor')
const highlightLine = require('./highlight-line')
const highlightWord = require('./highlight-word')

module.exports = (options = {}) => {
	return (tree) => {
		visit(tree, 'element', visitor)
	}

	function visitor(node, index, parentNode) {
		if (parentNode.tagName === 'pre' && node.tagName === 'code') {
			// syntax highlight
			const lang = node.properties.className ? node.properties.className[0].split('-')[1] : 'md'
			let result = refractor.highlight(nodeToString(node), lang)

			// line highlight
			const linesToHighlight = rangeParser(node.properties.line || '0')
			result = highlightLine(result, linesToHighlight)

			// word highlight
			result = highlightWord(result)

			node.children = result
		}
	}
}
```

How do I get access to `theme` & `showLineNumbers` directly in `pre` tag?

Minimal repro is here → https://github.com/deadcoder0904/better-code-blocks

My code is based on this amazing blog post → https://ped.ro/blog/code-blocks-but-better

The code for the blog post is here → https://github.com/peduarte/ped.ro

The only difference is I am using `@mdxjs/loader` with Tailwind & he is using `mdx-bundler` with Stitches :)

**Comment by wooorm** [comment]
I believe this is an xdm related question, and not necessarily related to rehype/html.
How to access meta as props is described in xdm’s readme: https://github.com/wooorm/xdm#syntax-highlighting-with-the-meta-field

**Reply by wooorm** [reply]
From your code: it looks like these props are passed to the `code`, not the `pre`, maybe you can pass a `code` component instead with `mdx-js/mdx`

**Reply by deadcoder0904** [reply]
Yes, the blog post I linked to → https://github.com/peduarte/ped.ro uses `mdx-bundler` & also uses `xdm`. But I don't. I use `mdx-loader`.

> From your code: it looks like these props are passed to the code, not the pre, maybe you can pass a code component instead with mdx-js/mdx

Yes, I can access them using `code` component but I need to use `pre` for styling purposes. Also, doesn't HTML's best practices say use `code` for `inline-code` & `pre>code` for `block-code` → https://stackoverflow.com/a/4611735/6141587

Where is the `mdx-js/mdx`? In the `next.config.js`? That would miss the whole styles, no?

Can't I just access `theme` & `showLineNumbers` directly? I did try `if ((node.tagName === 'pre' || node.tagName === 'code') && node.data && node.data.meta)` in `meta-attribute.js` but it didn't work but I feel like `rehype` plugin is where the error is. Am I right about that?

**Reply by wooorm** [reply]
`mdx-js/mdx` has some weird things going on in several places but especially in my opinion for how it deals with `meta`. You’re telling me it’s weird, which I agree with, but can’t do anything with. I made xdm to do several things, in my opinion, much clearer.

> doesn't HTML's best practices say use code for inline-code & pre>code for block-code → stackoverflow.com/a/4611735/6141587

`code`, the HTML element, can exist in many places. In a heading, in a pre, in an article. If you want to go into HTML semantics, then let the `pre` do what it’s supposed to do (be a block), and let the code do what it’s supposed to do (sometimes have a `className="language-js"`, and sometimes not).
Your theme field is similar to that language class: maybe handle code as a component, and branch to do different things based on whether a theme is set or not.

> Where is the mdx-js/mdx? In the next.config.js? That would miss the whole styles, no?

I am sorry, I don’t understand what you mean here?

> Can't I just access theme & showLineNumbers directly? I did try if ((node.tagName === 'pre' || node.tagName === 'code') && node.data && node.data.meta) in meta-attribute.js but it didn't work but I feel like rehype plugin is where the error is. Am I right about that?

`node.data.meta` exists on *code*. If you want to have it on `pre`, you need to look for `pre` only, then look through its children, find a child `code` with a `data.meta`, and use that to change the `pre`.

**Reply by wooorm** [reply]
You might be better of reading through a couple of articles on how to traverse trees: https://unifiedjs.com/learn/

**Reply by deadcoder0904** [reply]
@wooorm I've already read it a couple of months back.

> I am sorry, I don’t understand what you mean here?

You said: `From your code: it looks like these props are passed to the code, not the pre, maybe you can pass a code component instead with mdx-js/mdx` so I asked where should I use `code` with `mdx-js/mdx`? I think what you mean here is passing `code` instead of `pre` to `MDXComponents`, right?

> node.data.meta exists on code.

What do you mean it exists on `code` only? Do you mean `meta-attribute.js` file? I did change it to use `pre` & `code`, check the `if` statement:

```js
const visit = require('unist-util-visit')

var re = /\b([-\w]+)(?:=(?:"([^"]*)"|'([^']*)'|([^"'\s]+)))?/g

module.exports = (options = {}) => {
	return (tree) => {
		visit(tree, 'element', visitor)
	}

	function visitor(node, index, parentNode) {
		var match

		if ((node.tagName === 'pre' || node.tagName === 'code') && node.data && node.data.meta) {
			re.lastIndex = 0 // Reset regex.

			while ((match = re.exec(node.data.meta))) {
				node.properties[match[1]] = match[2] || match[3] || match[4] || ''
				parentNode.properties[match[1]] = match[2] || match[3] || match[4] || ''
			}
		}
	}
}
```

> If you want to have it on pre, you need to look for pre only, then look through its children, find a child code with a data.meta, and use that to change the pre.

Ha, this I understood. I need to write another traversal in `meta-attribute.js`, right?

**Reply by wooorm** [reply]
> I think what you mean here is passing code instead of pre to MDXComponents, right?

Correct

---

`node.data.meta` exists on `code`, that means that `if (node.tagName === 'code' && node.data && node.data.meta) {` does work, but `if (node.tagName === 'pre' && node.data && node.data.meta) {` **never matches**.

That being said, the following should work:

```js
		if (node.tagName === 'code' && node.data && node.data.meta) {
			re.lastIndex = 0 // Reset regex.

			while ((match = re.exec(node.data.meta))) {
				parentNode.properties[match[1]] = match[2] || match[3] || match[4] || ''
			}
		}
```

This is close to what you had, and you did it quite smartly. I would’ve done it the other way around:

```js
		if (node.tagName === 'pre') {
			const code = node.children.find(d => d.tagName === 'code')
			if (code.data && code.data.meta) {
				re.lastIndex = 0 // Reset regex.

				while ((match = re.exec(code.data.meta))) {
					node.properties[match[1]] = match[2] || match[3] || match[4] || ''
				}
			}
		}
```

Both *should* work

**Reply by deadcoder0904** [reply]
@wooorm I tested different variations with `MDXComponents` & tried using `pre`, `code`, `pre.code` & got too much confusion for the last 4 days but both the above solutions aren't working at all.

I want to do exactly what you said:

> `node.data.meta` exists on `code`. If you want to have it on `pre`, you need to look for `pre` only, then look through its children, find a child code with a `data.meta`, and use that to change the `pre`.

I changed your above code to have `console.log` but it never logs the `console.log('inside')`

# 1st solution

```ts
const visit = require('unist-util-visit')

var re = /\b([-\w]+)(?:=(?:"([^"]*)"|'([^']*)'|([^"'\s]+)))?/g

module.exports = (options = {}) => {
	return (tree) => {
		visit(tree, 'element', visitor)
	}

	function visitor(node, index, parentNode) {
		let match
		console.log('meta-attribute.js')
		if (node.tagName === 'code' && node.data && node.data.meta) {
			re.lastIndex = 0 // Reset regex.
			console.log('inside')
			while ((match = re.exec(node.data.meta))) {
				parentNode.properties[match[1]] = match[2] || match[3] || match[4] || ''
			}
		}
	}
}
```

It only logs `meta-attribute.js` multiple times but never logs `inside`

# 2nd solution

```ts
const visit = require('unist-util-visit')

var re = /\b([-\w]+)(?:=(?:"([^"]*)"|'([^']*)'|([^"'\s]+)))?/g

module.exports = (options = {}) => {
	return (tree) => {
		visit(tree, 'element', visitor)
	}

	function visitor(node, index, parentNode) {
		let match
		console.log('meta-attribute.js')
		if (node.tagName === 'pre') {
			const code = node.children.find((d) => d.tagName === 'code')
			if (code.data && code.data.meta) {
				console.log('inside')
				re.lastIndex = 0 // Reset regex.

				while ((match = re.exec(code.data.meta))) {
					console.log(node.properties, match)
					node.properties[match[1]] = match[2] || match[3] || match[4] || ''
				}
			}
		}
	}
}
```

It doesn't log `inside` here either.

My MDXComponents is using `Pre` like:

```tsx
import React from 'react'

type PreProps = {
	children: React.ReactChild
	theme: string
	showLineNumbers: string
}

export const Pre = ({ children, theme, showLineNumbers }: PreProps) => {
	console.log({ theme, showLineNumbers })
	return (
		<pre
			className={`px-4 py-3 overflow-x-auto rounded-md ${
				theme ? `${theme}-theme` : 'bg-gray-900'
			} ${showLineNumbers ? 'line-numbers' : ''}`}
		>
			{children}
		</pre>
	)
}

export const MDXComponents = {
	pre: Pre
}
```

And `theme` & `showLineNumbers` come `undefined`.

All I want is `pre > code`.

```html
<pre class="theme showLineNumbers">
	<code class="language-jsx ">
		...
	</code>
</pre>
```

where `theme` & `showLineNumbers` are actual values that I pass in MDX.

````mdx
```mdx theme=turq showLineNumbers=true
console.log('Yay that works!')
```
````

I'll go through the docs again but let me know if you have an answer :)

**Reply by wooorm** [reply]
You can probably figure out what’s going on by adding a `console.log` statement in both your solutions:

For 1:

```diff
	function visitor(node, index, parentNode) {
		let match
-		console.log('meta-attribute.js')
+		console.log('meta-attribute.js', node)
```

For 2:

```diff
		console.log('meta-attribute.js')
		if (node.tagName === 'pre') {
			const code = node.children.find((d) => d.tagName === 'code')
+			console.log(code)
```

**Reply by wooorm** [reply]
That is to say, `mdx-js/mdx` passes different props.

**Reply by deadcoder0904** [reply]
@wooorm I already did that. I didn't find anything like `code.data` so that's why `inside` isn't logging to the console.

Is `code.data` passed through `xdm` so that might be why it's `undefined`?

I do have `metastring` under `properties` like:

```js
{
  type: 'element',
  tagName: 'code',
  properties: {
    className: [ 'language-html' ],
    metastring: 'line=2-4',
    line: '2-4'
  },
  children: [
    {
      type: 'text',
      value: '<div>\n' +
        '    <h1>Hello world</h1>\n' +
        '    <p>This is a code block example.</p>\n' +
        '    <p>Do you like it?</p>\n' +
        '</div>\n'
    }
  ],
  position: {
    start: { line: 112, column: 1, offset: 2394 },
    end: { line: 118, column: 4, offset: 2511 }
  }
}
```

Here's another using `theme` & `showLineNumbers:

```js
{
  type: 'element',
  tagName: 'code',
  properties: {
    className: [ 'language-jsx' ],
    metastring: 'theme=turq showLineNumbers=true',
    theme: 'turq',
    showLineNumbers: 'true'
  },
  children: [
    {
      type: 'text',
      value: "import React from 'react'\n" +
        '\n' +
        'export function Counter({ initialCount = 0 }) {\n' +
        '    const [count, setCount] = React.useState(initialCount)\n' +
        '    return (\n' +
        '        <button type="button" onClick={() => setCount((prevCount) => prevCount + 1)}>\n' +
        '            {count}\n' +
        '        </button>\n' +
        '    )\n' +
        '}\n'
    }
  ],
  position: {
    start: { line: 36, column: 1, offset: 911 },
    end: { line: 47, column: 4, offset: 1202 }
  }
}
```

Is this what I should read?

**Reply by wooorm** [reply]
You placed the console logs in other places. Please apply my suggestions to the letter.
You will see that mdx-js/mdx uses `.properties` on code.

edit: well, that meta string does look at lot like what you’re looking for?

**Reply by deadcoder0904** [reply]
@wooorm nope, it's in the correct places! I double-checked now.

Yeah, I think that's what I want. Just need to do some JS manipulation to get what I want :)

**Reply by deadcoder0904** [reply]
Maybe I need to use https://github.com/wooorm/xdm#syntax-highlighting-with-the-meta-field to make it easy & then your code might work haha :)

**Reply by deadcoder0904** [reply]
Bdw https://github.com/remcohaszing/remark-mdx-code-meta links wrong way here → https://github.com/wooorm/xdm#syntax-highlighting-with-the-meta-field

**Reply by wooorm** [reply]
Maybe good to post that as an issue, or fix it in a PR, there?

**Reply by deadcoder0904** [reply]
@wooorm I can but it isn't working haha. Probably needs `xdm` or something. I get this error:

> ./src/pages/tutorial/better-code-blocks/index.mdx
SyntaxError: JSX value should be either an expression or a quoted JSX text (1:11)

**Reply by wooorm** [reply]
Correct, it uses newer features that are currently only available in xdm

**Comment by deadcoder0904** [comment, marked as answer]
This did the trick:

# meta-attribute.js

```js
const visit = require('unist-util-visit')

var re = /\b([-\w]+)(?:=(?:"([^"]*)"|'([^']*)'|([^"'\s]+)))?/g

module.exports = (options = {}) => {
	return (tree) => {
		visit(tree, 'element', visitor)
	}

	function visitor(node, index, parentNode) {
		let match

		if (node.tagName === 'pre') {
			const code = node.children.find((d) => d.tagName === 'code')

			if (code.properties.metastring) {
				code.properties.metastring.split(' ').forEach((item) => {
					const [metaKey, metaValue] = item.split('=')
					node.properties[metaKey] = metaValue || ''
				})
			}
		}
	}
}
```

@wooorm Any way I can remove my custom properties like `theme` from DOM?

**Reply by wooorm** [reply]
If you put them in `properties`, which the above code is doing, they will be passed, which was your question, so I’m not quite sure how you want *both* to pass them to the DOM, and *not* to pass them too the DOM?

**Reply by deadcoder0904** [reply]
@wooorm I get console warning:

> Warning: React does not recognize the `showLineNumbers` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `showlinenumbers` instead. If you accidentally passed it from a parent component, remove it from the DOM element.

This is because my `code` in the DOM looks like:

```html
<code
  class="language-jsx"
  metastring="theme=turq showLineNumbers=true"
  theme="turq"
  showlinenumbers="true"
>
  ...
</code>
```

Other than `class`, I don't think the other properties should exist in the DOM.

This is probably happening because of `children` in my `Pre` component which renders `code`:

```tsx
import React from 'react'

type PreProps = {
	children: React.ReactChild
	theme: string
	showLineNumbers: string
}

export const Pre = ({ children, theme, showLineNumbers }: PreProps) => {
	return (
		<pre
			className={`px-4 py-3 overflow-x-auto rounded-md ${
				theme ? `${theme}-theme` : 'bg-gray-900'
			} ${showLineNumbers ? 'line-numbers' : ''}`}
		>
			{children}
		</pre>
	)
}
```

Idk how to extract those properties from `code` to not give the warning in the console & clean up the DOM. Is it possible?

**Reply by wooorm** [reply]
Yes, this is why I don’t like what `mdx-js/mdx` is doing. Again, I would suggest xdm.
There are two possible solutions: either pass a code component which ignores those fields, or remove those fields in your AST transform

**Reply by deadcoder0904** [reply]
@wooorm Can `@mdx-js/loader` replace `xdm` completely? And I just saw the README & it says its ESM only which Next.js doesn't support. So I'm using all your modules on previous major versions.

I guess it's not that big of a problem for now. And thank you so much for all the help :)

**Reply by wooorm** [reply]
xdm frequently works at build time. Other next users were able to use it: https://github.com/wooorm/xdm#next.
If your use case is more involved and you want to use xdm at runtime in next, either using 
https://github.com/wellcometrust/next-plugin-transpile-modules or manually using esbuild or webpack to turn it into cjs is always an option.
